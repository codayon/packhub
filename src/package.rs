use std::{str::FromStr, sync::Mutex};

use anyhow::Result;
use chrono::{DateTime, Utc};
use lenient_semver::parse;
use semver::Version;

#[derive(Debug, PartialEq, Clone)]
pub enum Dist {
    Ubuntu(Option<Version>),
    Debian(Option<Version>),
    Fedora(Option<Version>),
}

#[derive(Debug, PartialEq)]
pub enum Arch {
    Amd64,
}

impl FromStr for Arch {
    type Err = ();

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "amd64" => Ok(Arch::Amd64),
            _ => Err(()),
        }
    }
}

#[derive(Debug, PartialEq)]
enum Type {
    Deb,
    Rpm,
}

#[derive(Debug)]
pub struct Package {
    tipe: Type,
    pub(crate) dist: Option<Dist>,
    url: String,
    ver: String,
    data: Mutex<Option<Vec<u8>>>,
    created: DateTime<Utc>,
}

impl PartialEq for Package {
    fn eq(&self, other: &Self) -> bool {
        self.tipe == other.tipe
            && self.dist == other.dist
            && self.url == other.url
            && self.ver == other.ver
            && *self.data.lock().unwrap() == *other.data.lock().unwrap()
            && self.created == other.created
    }
}

struct DetectError;

impl Package {
    pub fn detect_package(
        name: &str,
        ver: String,
        url: String,
        created: DateTime<Utc>,
    ) -> Result<Package, ()> {
        // Split the extension first.
        // If we don't recognize it, then return error.
        let Some((tipe, splitted)) = split_extention(name) else {
            return Err(());
        };

        let mut dist: Option<Dist> = None;
        let sections: Vec<&str> = splitted.split(['-', '_']).collect();

        for section in sections {
            match section {
                dst if dst.contains("ubuntu") => dist = Some(Dist::Ubuntu(parse_version(dst))),
                dst if dst.contains("debian") => dist = Some(Dist::Debian(parse_version(dst))),
                dst if dst.contains("fedora") => dist = Some(Dist::Fedora(parse_version(dst))),
                _ => (),
            }
        }

        Ok(Package {
            tipe,
            dist,
            url,
            ver,
            data: Mutex::new(None),
            created,
        })
    }

    pub fn is_deb(&self) -> bool {
        self.tipe == Type::Deb
    }

    /// Check if the package is for Ubuntu
    pub fn for_ubuntu(&self) -> bool {
        matches!(self.dist, Some(Dist::Ubuntu(_)))
    }

    /// Return the distribution for which it was packaged
    pub fn distribution(&self) -> &Dist {
        self.dist.as_ref().unwrap()
    }

    /// Version of the package
    pub fn version(&self) -> &str {
        &self.ver
    }

    pub fn download_url(&self) -> &str {
        &self.url
    }

    pub fn file_name(&self) -> &str {
        &self.url.split('/').last().unwrap()
    }

    /// Download package data
    ///
    /// It is required to call this function before calling the `data()` function.
    pub async fn download(&self) -> Result<()> {
        let data = reqwest::get(self.download_url()).await?.bytes().await?;
        *self.data.lock().unwrap() = Some(data.to_vec());
        Ok(())
    }

    /// Return the data of the package.
    ///
    /// It is required to call the `download()` function before calling this.
    /// Otherwise, `None` is returned.
    pub fn data(&self) -> Option<Vec<u8>> {
        self.data.lock().unwrap().clone()
    }

    #[cfg(test)]
    /// Set the internal package data.
    ///
    /// It's for testing purpose.
    pub fn set_data(&self, data: Vec<u8>) {
        *self.data.lock().unwrap() = Some(data);
    }

    pub fn creation_date(&self) -> &DateTime<Utc> {
        &self.created
    }
}

/// Parses the version from the distribution identifier `dist`.
///
/// For instance, for a distribution identifier `ubuntu22.10` it will
/// parse the version as `22.10`.
fn parse_version(dist: &str) -> Option<Version> {
    parse(split_at_numeric(dist)?).ok()
}

/// Splits the string `s` at the first occurence of a numeric digit.
///
/// It is used to extract version number from strings, such as for "ubuntu24.10" it would
/// return "24.10".
fn split_at_numeric(s: &str) -> Option<&str> {
    for (curr, (index, next)) in s.chars().zip(s.char_indices().skip(1)) {
        if curr.is_ascii_alphabetic() && next.is_ascii_digit() {
            return Some(&s[index..]);
        }
    }

    None
}

fn split_extention(s: &str) -> Option<(Type, &str)> {
    let mut str = String::with_capacity(3);
    let mut index = 0;

    for (idx, ch) in s.char_indices().rev() {
        if ch == '.' {
            index = idx;
            break;
        } else {
            str.push(ch);
        }
    }

    if index == 0 {
        return None;
    }

    let splitted = &s[0..index];

    // `str` is in reverse order, so we try to match it reversely.
    let tipe = match str.as_str() {
        "bed" => Type::Deb,
        "mpr" => Type::Rpm,
        _ => return None,
    };

    Some((tipe, splitted))
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_package() {
        let pack = Package::detect_package(
            "OpenBangla-Keyboard_2.0.0-ubuntu22.04.deb",
            "2.0.0".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(pack.version(), "2.0.0");
        assert_eq!(pack.dist, Some(Dist::Ubuntu(Some(parse("22.04").unwrap()))));
        assert_eq!(pack.tipe, Type::Deb);

        let pack = Package::detect_package(
            "OpenBangla-Keyboard_2.0.0-fedora36.rpm",
            "2.0.0".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(pack.version(), "2.0.0");
        assert_eq!(pack.dist, Some(Dist::Fedora(Some(parse("36").unwrap()))));
        assert_eq!(pack.tipe, Type::Rpm);

        let pack = Package::detect_package(
            "caprine_2.56.1_amd64.deb",
            "v2.56.1".to_owned(),
            String::new(),
            DateTime::UNIX_EPOCH,
        )
        .unwrap();
        assert_eq!(pack.version(), "v2.56.1");
        assert_eq!(pack.dist, None);
        assert_eq!(pack.tipe, Type::Deb);
    }

    #[test]
    fn test_split_extension() {
        assert_eq!(
            split_extention("OpenBangla-Keyboard_2.0.0-ubuntu22.04.deb"),
            Some((Type::Deb, "OpenBangla-Keyboard_2.0.0-ubuntu22.04"))
        );
        assert_eq!(
            split_extention("OpenBangla-Keyboard_2.0.0-fedora36.rpm"),
            Some((Type::Rpm, "OpenBangla-Keyboard_2.0.0-fedora36"))
        );
        assert_eq!(split_extention("caprine_2.56.1_amd64.snap"), None);
        assert_eq!(split_extention("deb"), None);
    }

    #[test]
    fn test_split_test() {
        assert_eq!(split_at_numeric("ubuntu24.10"), Some("24.10"));
        assert_eq!(split_at_numeric("ubuntu"), None);
    }

    #[test]
    fn test_parse_version() {
        assert_eq!(
            parse_version("ubuntu22.10").unwrap(),
            Version::new(22, 10, 0)
        );
        assert_eq!(parse_version("fedora37").unwrap(), Version::new(37, 0, 0));
    }
}
