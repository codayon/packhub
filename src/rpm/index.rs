use askama::Template;

use super::package::{Dependency, RPMPackage};

#[derive(Debug, Template)]
#[template(path = "primary.xml")]
struct Primary {
    packages: Vec<RPMPackage>,
}

#[derive(Debug, Template)]
#[template(path = "filelists.xml")]
struct FileLists {
    packages: Vec<RPMPackage>,
}

#[derive(Debug, Template)]
#[template(path = "other.xml")]
struct Other {
    packages: Vec<RPMPackage>,
}

pub fn get_primary_index(packages: Vec<RPMPackage>) -> String {
    let primary = Primary { packages };
    primary.render().unwrap()
}

pub fn get_filelists_index(packages: Vec<RPMPackage>) -> String {
    let list = FileLists { packages };
    list.render().unwrap()
}

pub fn get_other_index(packages: Vec<RPMPackage>) -> String {
    let list = Other { packages };
    list.render().unwrap()
}

#[cfg(test)]
mod tests {
    use std::fs::read;

    use chrono::DateTime;
    use insta::assert_snapshot;

    use crate::package::Package;

    use super::*;

    #[test]
    fn test_rpm_indices() {
        let package = Package::detect_package("OpenBangla-Keyboard_2.0.0-fedora38.rpm", "2.0.0".to_owned(), "https://github.com/OpenBangla/OpenBangla-Keyboard/releases/download/2.0.0/OpenBangla-Keyboard_2.0.0-fedora38.rpm".to_owned(), DateTime::UNIX_EPOCH).unwrap();
        let data = read("data/OpenBangla-Keyboard_2.0.0-fedora38.rpm").unwrap();
        package.set_data(data);
        let package = RPMPackage::from_package(&package).unwrap();
        let packages = vec![package];

        assert_snapshot!(get_primary_index(packages.clone()));

        assert_snapshot!(get_filelists_index(packages.clone()));

        assert_snapshot!(get_other_index(packages));
    }
}
