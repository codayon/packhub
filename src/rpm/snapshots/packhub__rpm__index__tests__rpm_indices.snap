---
source: src/rpm/index.rs
expression: get_primary_index(packages.clone())
---
<?xml version="1.0" encoding="UTF-8"?>
<metadata xmlns="http://linux.duke.edu/metadata/common" xmlns:rpm="http://linux.duke.edu/metadata/rpm" packages="1">

<package type="rpm">
  <name>openbangla-keyboard</name>
  <arch>x86_64</arch>
  <version epoch="0" ver="2.0.0" rel="1.fc38"/>
  <checksum type="sha256" pkgid="YES">4ea19432c75d1b7ddc6dc2adbb10301fea142913834decd0455e1249cbaf3c14</checksum>
  <summary>OpenSource Bengali input method</summary>
  <description>OpenBangla Keyboard is an OpenSource, Unicode compliant Bengali Input Method for GNU/Linux systems. It&#x27;s a full fledged Bengali input method with typing automation tools, includes many famous typing methods such as Avro Phonetic, Probhat, Munir Optima, National (Jatiya) etc.

Most of the features of Avro Keyboard are present in OpenBangla Keyboard. So Avro Keyboard users will feel at home with OpenBangla Keyboard in Linux.</description>
  <packager></packager>
  <url>https://openbangla.github.io/</url>
  <time file="0" build="1684606381"/>
  <size package="4615096" installed="12106690" archive="12112024"/>
  <location href="package/2.0.0/OpenBangla-Keyboard_2.0.0-fedora38.rpm"/>
  <format>
    <rpm:license>GPLv3</rpm:license>
    <rpm:vendor>OpenBangla Team</rpm:vendor>
    <rpm:group>unknown</rpm:group>
    <rpm:buildhost>3aebfc71f602</rpm:buildhost>
    <rpm:sourcerpm>openbangla-keyboard-2.0.0-1.fc38.src.rpm</rpm:sourcerpm>
    <rpm:header-range start="4504" end="14821"/>
    <rpm:provides>
    
      
      <rpm:entry name="application()"/>
      
    
      
      <rpm:entry name="application(openbangla-keyboard.desktop)"/>
      
    
      
      <rpm:entry name="config(openbangla-keyboard)" flags="EQ" epoch="0" ver="2.0.0-1.fc38" rel="1.fc38"/>
      
    
      
      <rpm:entry name="metainfo()"/>
      
    
      
      <rpm:entry name="metainfo(io.github.openbangla.keyboard.metainfo.xml)"/>
      
    
      
      <rpm:entry name="openbangla-keyboard" flags="EQ" epoch="0" ver="2.0.0-1.fc38" rel="1.fc38"/>
      
    
      
      <rpm:entry name="openbangla-keyboard(x86-64)" flags="EQ" epoch="0" ver="2.0.0-1.fc38" rel="1.fc38"/>
      
    
    </rpm:provides>
    <rpm:requires>
    
      
      <rpm:entry name="ibus" flags="GE" epoch="0" ver="1.5.1"/>
      
    
      
      <rpm:entry name="ibus-libs" flags="GE" epoch="0" ver="1.5.1"/>
      
    
      
      <rpm:entry name="ld-linux-x86-64.so.2()(64bit)"/>
      
    
      
      <rpm:entry name="ld-linux-x86-64.so.2(GLIBC_2.3)(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Core.so.5()(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Core.so.5(Qt_5)(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Core.so.5(Qt_5.15)(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Gui.so.5()(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Gui.so.5(Qt_5)(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Network.so.5()(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Network.so.5(Qt_5)(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Widgets.so.5()(64bit)"/>
      
    
      
      <rpm:entry name="libQt5Widgets.so.5(Qt_5)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.10)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.14)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.15)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.17)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.18)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.2.5)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.25)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.27)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.28)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.29)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.3)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.3.4)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.32)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.33)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.34)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.4)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.5)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.6)(64bit)"/>
      
    
      
      <rpm:entry name="libc.so.6(GLIBC_2.9)(64bit)"/>
      
    
      
      <rpm:entry name="libgcc_s.so.1()(64bit)"/>
      
    
      
      <rpm:entry name="libgcc_s.so.1(GCC_3.0)(64bit)"/>
      
    
      
      <rpm:entry name="libgcc_s.so.1(GCC_3.3)(64bit)"/>
      
    
      
      <rpm:entry name="libgcc_s.so.1(GCC_4.2.0)(64bit)"/>
      
    
      
      <rpm:entry name="libgio-2.0.so.0()(64bit)"/>
      
    
      
      <rpm:entry name="libglib-2.0.so.0()(64bit)"/>
      
    
      
      <rpm:entry name="libgobject-2.0.so.0()(64bit)"/>
      
    
      
      <rpm:entry name="libibus-1.0.so.5()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libm.so.6(GLIBC_2.2.5)(64bit)"/>
      
    
      
      <rpm:entry name="libstdc++.so.6()(64bit)"/>
      
    
      
      <rpm:entry name="libstdc++.so.6(CXXABI_1.3)(64bit)"/>
      
    
      
      <rpm:entry name="libstdc++.so.6(CXXABI_1.3.1)(64bit)"/>
      
    
      
      <rpm:entry name="libstdc++.so.6(GLIBCXX_3.4)(64bit)"/>
      
    
      
      <rpm:entry name="libstdc++.so.6(GLIBCXX_3.4.21)(64bit)"/>
      
    
      
      <rpm:entry name="libzstd" flags="GE" epoch="0" ver="1.3.3"/>
      
    
      
      <rpm:entry name="libzstd.so.1()(64bit)"/>
      
    
      
      <rpm:entry name="qt5-qtbase" flags="GE" epoch="0" ver="5.9.0"/>
      
    
      
      <rpm:entry name="rtld(GNU_HASH)"/>
      
    
    </rpm:requires>
  </format>
</package>

</metadata>
